### PR TITLE
[FLINK-22317][table] Support DROP column/constraint/watermark for ALTER TABLE statement

### PR DESCRIPTION
**Pull request by LadyForest**
## What is the purpose of the change

This pull request aims to support `ALTER TABLE table_identifier DROP` statement.

## Brief change log

This PR contains two commits
  - 4b00202d implements the syntax
  - 93b32c4a supports operation conversion

## Verifying this change

This change added tests and can be verified as follows:

- `SqlToOperationConverterTest`
- `TableEnvironmentTest`
- `table.q`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? FLINK-22320 will document this new feature

### CODE REVIEW

**Review by fsk119 — COMMENTED** [review]
Thanks for your great work. I left some comments.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/AlterSchemaConverter.java [hunk 1]
```diff
@@ -115,77 +120,135 @@ public Schema applySchemaChange(SqlAlterTableSchema alterTableSchema, Schema ori
...
+        Schema.Builder schemaBuilder = Schema.newBuilder();
+        buildUpdatedColumn(
+                schemaBuilder,
+                originTable,
+                (builder, column) -> builder.fromColumns(Collections.singletonList(column)));
+        buildUpdatedWatermark(schemaBuilder, originTable);
+        return schemaBuilder.build();
+    }
+
+    public Schema applySchemaChange(ContextResolvedTable originTable) {
```
[the remark this answers is not in the document]
Comment (LadyForest) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/AlterSchemaConverter.java:238]: > It's hard for other to understand what's the usage of this method from the signature. How about
> 
> ```
> applySchemaChange(SqlAlterWatermark, ContextResolvedTable)
> ```

I agree that the method is not straightforward, but `SqlAlterWatermark` is not used. How about adding the method description?
```java
/** Convert ALTER TABLE DROP WATERMARK to generate an updated {@link Schema}. */
applySchemaChange(ContextResolvedTable)
```

[the remark this answers is not in the document]
Comment (fsk119) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/AlterSchemaConverter.java:238]: The java doc is not as straightforward as the signature. It's fine to leave something unused. In the `SqlToOperationConverter#convertBeginStatementSet`, we also use the `SqlBeginStatementSet` as the input parameter that is not used.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/AlterSchemaConverter.java [hunk 2]
```diff
@@ -604,24 +703,90 @@ private void validateColumnName(
...
                                 throw new ValidationException(
                                         String.format(
-                                                "Old column %s is referred by computed column %s, currently doesn't "
-                                                        + "allow to rename column which is referred by computed column.",
-                                                originColumnName,
+                                                "%sThe column `%s` is referenced by computed column %s.",
+                                                EX_MSG_PREFIX,
+                                                columnToAlter,
                                                 computedColumn.asSummaryString()));
                             }
```
[the remark this answers is not in the document]
Comment (LadyForest) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/AlterSchemaConverter.java:713]: > BTW, the drop order will influence the logic here, e.g. drop physical column first and then drop the computed column using the physical column?

I think `ALTER TABLE T DROP (col1, col2, ...)` is a bulk operation, so the order should not influence the result. As long as the computed column is dropped together with the physical column, it should be a valid operation. WDYT?

**Comment by fsk119** [comment]
Thanks for the update! I add a commit to improve some codes. Please take a look at the last commit when you are free.

**Comment by LadyForest** [comment]
> Thanks for the update! I add a commit to improve some codes. Please take a look at the last commit when you are free.

Thanks for the improvement, LGTM 👍

**Pull request closed without merging**